Add GSoC 2026 blog post for "Introducing Float80 and Float128 in LLVM libc" (#92)
llvm/llvm-blog-www: content/posts/2026-08-22-gsoc-float80-and-float128-in-llvm-libc.md

Reviewer: Nit: formatting.

```suggestion
```sh
```

---
author: "Sukumar Sawant"
date: "2026-08-22"
tags: ["GSoC", "libc", "math", "floating-point"]

> title: "GSoC 2026: Float80 and Float128 Emulation in LLVM libc"

Reviewer: The title should probably follow the project name as it is on the GSoC website as closely as possible.

```suggestion
title: "GSoC 2026: Enable float80/float128 on Unsupported Targets for LLVM libc"
```

---

## Introduction

Hi, my name is Sukumar Sawant and I had the pleasure of working on the project on introducing `float80` and `float128` (emulated) in LLVM libc this Google Summer of Code 2026 along with my mentors Tue Ly, Nicolas Celik and Krishna Pandey.

## "Why"?

> `float80` and `float128` are the two extended-precision floating-point formats LLVM libc relies on. `float80` is the 80-bit x87 extended precision format, with 1 sign bit, 15 exponent bits, and a 64-bit significand carrying an explicit integer bit. `float128` is the IEEE 754 `binary128` ("quad precision") format, with 1 sign bit, 15 exponent bits, and 112 stored mantissa bits.

Reviewer: ```suggestion
`float80` and `float128` are the two extended-precision floating-point formats LLVM libc supports. `float80` is the 80-bit x87 extended precision format, with 1 sign bit, 15 exponent bits, and a 64-bit significand carrying an explicit integer bit. `float128` is the IEEE 754 `binary128` ("quad precision") format, with 1 sign bit, 15 exponent bits, and 112 stored mantissa bits.
```

* Clang on certain non-x86 targets may not enable `__float128` by default.

This matters because LLVM libc needs to build and test its high-precision math routines consistently across every target it supports, not just the ones where these types happen to exist natively. Where the compiler provides no native type, the templated implementations cannot be compiled at all.

The goal of this project was to lift that restriction by emulating both formats in software, so that the same templated implementations build and run everywhere, producing identical results whether the underlying type is native or emulated.

The exact goals of this project were to:

1. Add software emulation classes for the `float80` and `float128` formats.
2. Provide the operator overloads and conversions needed for these classes to be used in place of the native types.
3. Integrate the emulated types into the existing floating-point support infrastructure.
4. Modify the existing `float128` math functions to work correctly with the emulated type.

## Interesting Challenges

### 1. A `bit_cast` failure that only showed on Windows

> `Float128` stores its value in a `UInt128`, which on Windows, without a native 128-bit integer, is `BigInt`. Everything that goes through `cpp::bit_cast` requires it to be trivially constructible and trivially copyable, which was failing in this case.

Reviewer: ```suggestion
`Float128` stores its value in a `UInt128`, which on Windows, without a native 128-bit integer type, is `BigInt`. LLVM-libc's `cpp::bit_cast` requires the target type to be trivially constructible and trivially copyable, which was failing in this case.
```

> `Float128` stores its value in a `UInt128`, which on Windows, without a native 128-bit integer, is `BigInt`. Everything that goes through `cpp::bit_cast` requires it to be trivially constructible and trivially copyable, which was failing in this case.

Reviewer: > Everything that goes through `cpp::bit_cast` requires it to be trivially constructible and trivially copyable, which was failing in this case.

What is "it" (in "requires it") referring to?

> `Float128` stores its value in a `UInt128`, which on Windows, without a native 128-bit integer, is `BigInt`. Everything that goes through `cpp::bit_cast` requires it to be trivially constructible and trivially copyable, which was failing in this case.

Reviewer: "A `bit_cast` failure" and "which was failing in this case" sounds ambiguous. You should explain that it was a compile error:

```
D:\a\llvm-project\llvm-project\libc\src/__support/FPUtil\FPBits.h(863,12): error: no matching function for call to 'bit_cast'
  863 |     return cpp::bit_cast<T>(UP::bits);
      |            ^~~~~~~~~~~~~~~~
...
D:\a\llvm-project\llvm-project\libc\test\src\math\smoke\float128_test.cpp(21,12): note: in instantiation of function template specialization '__llvm_libc_23_0_0_git::fputil::Float128::Float128<double>' requested here
   21 |   Float128 a(1.0), b(1.0), c(2.0);
      |            ^
```

```cpp
cpp::is_trivially_constructible<To>::value && cpp::is_trivially_copyable<To>::value

> ```

Reviewer: I think it'd be easier to understand if you copied the whole `bit_cast` template:

```cpp
// This implementation of bit_cast requires trivially-constructible To, to avoid
// UB in the implementation.
template <typename To, typename From>
LIBC_INLINE static constexpr cpp::enable_if_t<
    (sizeof(To) == sizeof(From)) &&
        cpp::is_trivially_constructible<To>::value &&
        cpp::is_trivially_copyable<To>::value &&
        cpp::is_trivially_copyable<From>::value,
    To>
bit_cast(const From &from) {
  // ...
}
```

> The culprit was `BigInt` being zero-initialized in many places (in `libc/src/__support/big_int.h`).

Reviewer: It's one zero-initialization that was the problem:

```cpp
template <size_t Bits, bool Signed, typename WordType = uint64_t>
struct BigInt {
  // ...
  cpp::array<WordType, WORD_COUNT> val{}; // zero initialized.
  // ...
};
```

We fixed the issue by zero-initializing `BigInt` variables in many other places instead of zero-initializing `val` in the class definition. You explained it correctly below:

> The fix was to drop the zero-init from the declaration and push it into the constructors that actually need it

The default member initializer was enough to make the condition fail.

The fix was to drop the zero-init from the declaration and push it into the constructors that actually need it
([#206277](https://github.com/llvm/llvm-project/pull/206277)).

With the member initializer gone, the defaulted constructor is genuinely trivial, both traits become true, and `bit_cast` accepts the type, which fixed the Windows build.

> ### 2. The persistent ABI issue with `CORE-MATH`

Reviewer: Nit:

```suggestion
### 2. The persistent ABI issue with CORE-MATH
```

Also what was "persistent" about the issue? I guess it's an LLM-ism.

And the ABI issue isn't specific to CORE-MATH; it's about the native `__bf16` type in general.

Author: So this was a pre-GSoC issue with bfloat16, except causing unexpected FP exceptions(https://github.com/llvm/llvm-project/issues/186483).
Agreed, it's not really specific to core-math.

Persistent was in the sense that when the issue was opened previously I did go through many iterations fixing it for bf16, even after the pre-post rounding flags were added for CORE-MATH and it identified by tue as an ABI mismatch issue. This was first implemented in f128 and then verified for bf16 function last month as you know.

Though the title could be replaced with "The ABI issue with emulated types" instead if thats the case?

Reviewer: > Though the title could be replaced with "The ABI issue with emulated types" instead if thats the case?

Sure.

A native `float128` uses different registers than the emulated types like `Float128`, which under the hood uses a `UInt128` container.

> When comparing against other projects such as CORE-MATH, the `bfloat16` functions consistently failed.

Reviewer: ```suggestion
While testing LLVM-libc with the CORE-MATH project's test suite, the `bfloat16` (also emulated) functions consistently failed.
```

> So we tested our `float128` functions with CORE-MATH and modified them to use the emulation internally for calculations, and the native `float128` when available for input/output, so that in the end the registers used for I/O remain the same and the ABI mismatch is prevented.

Reviewer: Nit: I don't think it's common to refer to argument/return value passing as I/O.

```suggestion
So we tested our `float128` functions with CORE-MATH and modified them to use the emulation internally for calculations, and the native `float128` when available for input/output, so that in the end the registers used for parameters/returns remain the same and the ABI mismatch is prevented.
```

| `fmaximum_mag_numf128`, `fminimum_mag_numf128`                                                           | [#217549](https://github.com/llvm/llvm-project/pull/217549)     |
| `fdimf128`                                                                                               | [#217567](https://github.com/llvm/llvm-project/pull/217567)     |
| `iscanonicalf128`, `isnanf128`, `issignalingf128`                                                       | [#217134](https://github.com/llvm/llvm-project/pull/217134)     |
| `nextafterf128`, `nextupf128`, `nextdownf128`                                                           | [#217575](https://github.com/llvm/llvm-project/pull/217575)     |
| `fmodf128`, `remainderf128`                                                                              | [#217578](https://github.com/llvm/llvm-project/pull/217578)     |
| `sqrtf128`                                                                                               | [#216500](https://github.com/llvm/llvm-project/pull/216500)     |

> | `atan2f128`                                                                                              | [#216508](https://github.com/llvm/llvm-project/pull/216508)     |

Reviewer: Nit: inconsistent formatting (spaces).

Author: I was trying to see if I could extend the table to the full width of the page, but that didn't really work out. Missed to remove later.
Updated now , thanks!

  863 |     return cpp::bit_cast<T>(UP::bits);
      |            ^~~~~~~~~~~~~~~~
...
D:\a\llvm-project\llvm-project\libc\test\src\math\smoke\float128_test.cpp(21,12): note: in instantiation of function template specialization '__llvm_libc_23_0_0_git::fputil::Float128::Float128<double>' requested here
   21 |   Float128 a(1.0), b(1.0), c(2.0);
      |            ^
```

The culprit was `BigInt` being zero-initialized in `libc/src/__support/big_int.h`.

```
template <size_t Bits, bool Signed, typename WordType = uint64_t>
struct BigInt {
  // ...
  cpp::array<WordType, WORD_COUNT> val{}; // zero initialized.
  // ...
};
```

The default member initializer was enough to make the condition fail.

The fix was to drop the zero initialization from the declaration and push it into the constructors that actually need it
([#206277](https://github.com/llvm/llvm-project/pull/206277)).

With the member initializer gone, the defaulted constructor is genuinely trivial, both traits become true, and `bit_cast` accepts the type, which fixed the Windows build.

### 2. The ABI issue with emulated types

> A native `float128` uses different registers than the emulated types like `Float128`, which under the hood uses a `UInt128` container.

Reviewer 2: We should talk about `bfloat16` here.  That when the compiler support `bfloat16` under the name `__bf16`, it uses different registers than the emulation `BFloat16` type previously implemented in LLVM libc.

Reviewer 2: So it was kind of the opposite problem compared to Float128.  And the solutions converges for both `float128` and `bfloat16`.

| `fmaximum_mag_numf128`, `fminimum_mag_numf128`             | [#217549](https://github.com/llvm/llvm-project/pull/217549) |
| `fdimf128`                                                 | [#217567](https://github.com/llvm/llvm-project/pull/217567) |
| `iscanonicalf128`, `isnanf128`, `issignalingf128`          | [#217134](https://github.com/llvm/llvm-project/pull/217134) |
| `nextafterf128`, `nextupf128`, `nextdownf128`              | [#217575](https://github.com/llvm/llvm-project/pull/217575) |
| `fmodf128`, `remainderf128`                                | [#217578](https://github.com/llvm/llvm-project/pull/217578) |
| `sqrtf128`                                                 | [#216500](https://github.com/llvm/llvm-project/pull/216500) |
| `atan2f128`                                                | [#216508](https://github.com/llvm/llvm-project/pull/216508) |

## Conclusion

- We would now be able to use, test and build the `Float128` and `Float80` types even on platforms where they are not natively available.
- The `float128` functions are now enabled on targets like Windows with MSVC, macOS, and GPU targets.

> - We wil soon also be able to test `bfloat16` functions with core-math test suite.

Reviewer 3: can be removed

```cpp
// This implementation of bit_cast requires trivially-constructible To, to avoid
// UB in the implementation.
template <typename To, typename From>
LIBC_INLINE static constexpr cpp::enable_if_t<
    (sizeof(To) == sizeof(From)) &&
        cpp::is_trivially_constructible<To>::value &&
        cpp::is_trivially_copyable<To>::value &&
        cpp::is_trivially_copyable<From>::value,
    To>
bit_cast(const From &from) {
  // ...
}
```
``` sh
D:\a\llvm-project\llvm-project\libc\src/__support/FPUtil\FPBits.h(863,12): error: no matching function for call to 'bit_cast'
  863 |     return cpp::bit_cast<T>(UP::bits);
      |            ^~~~~~~~~~~~~~~~
...
D:\a\llvm-project\llvm-project\libc\test\src\math\smoke\float128_test.cpp(21,12): note: in instantiation of function template specialization '__llvm_libc_23_0_0_git::fputil::Float128::Float128<double>' requested here
   21 |   Float128 a(1.0), b(1.0), c(2.0);
      |            ^
```

The culprit was `BigInt` being zero-initialized in `libc/src/__support/big_int.h`.

> ```

Reviewer: ```suggestion
```cpp
```

The culprit was `BigInt` being zero-initialized in `libc/src/__support/big_int.h`.

> ```CPP

Reviewer: Nit: formatting.

```suggestion
```cpp
```

### 2. The ABI issue with emulated types

A native `float128` uses different registers than the emulated types like `Float128`, which under the hood uses a `UInt128` container.

> While testing LLVM-libc with the CORE-MATH project's test suite, the `bfloat16` (also emulated) functions consistently failed. This was because when compiler support bfloat16 under the name __bf16 is available, it uses different registers than the emulation [BFloat16](https://github.com/llvm/llvm-project/pull/144463) type previously implemented in LLVM libc .

Reviewer: ```suggestion
While testing LLVM-libc with the CORE-MATH project's test suite, the `bfloat16` (also emulated) functions consistently failed. This was because CORE-MATH uses the native `__bf16` type provided by the compiler, which uses different registers than the emulated [BFloat16](https://github.com/llvm/llvm-project/pull/144463) type previously implemented in LLVM libc.
```

| `fmaximum_mag_numf128`, `fminimum_mag_numf128`             | [#217549](https://github.com/llvm/llvm-project/pull/217549) |
| `fdimf128`                                                 | [#217567](https://github.com/llvm/llvm-project/pull/217567) |
| `iscanonicalf128`, `isnanf128`, `issignalingf128`          | [#217134](https://github.com/llvm/llvm-project/pull/217134) |
| `nextafterf128`, `nextupf128`, `nextdownf128`              | [#217575](https://github.com/llvm/llvm-project/pull/217575) |
| `fmodf128`, `remainderf128`                                | [#217578](https://github.com/llvm/llvm-project/pull/217578) |
| `sqrtf128`                                                 | [#216500](https://github.com/llvm/llvm-project/pull/216500) |
| `atan2f128`                                                | [#216508](https://github.com/llvm/llvm-project/pull/216508) |

## Conclusion

- We would now be able to use, test and build the `Float128` and `Float80` types even on platforms where they are not natively available.
- The `float128` functions are now enabled on targets like Windows with MSVC, macOS, and GPU targets.

> - The `Float128` functions can now also be tested with `CORE-MATH` test-suite to verify correctness.

Reviewer: Either say that *emulated* `float128` can be tested with the CORE-MATH test suite (was already the case for native), or that we will soon be able to test `bfloat16` functions with the CORE-MATH test suite.
